spec.html

```diff
@@ -7649,58 +7649,42 @@ <h1>Jobs and Job Queues</h1>
-    </emu-note>
-    <p>The following abstract operations are used to create and manage Jobs and Job Queues:</p>
+    <p>Jobs are scheduled for execution by ECMAScript host environments with abstract operations such as HostEnqueuePromiseJob. Such operations accept a PendingJob Record _pending_ as a parameter. These abstract operations schedule the PendingJob to be performed at some future time. Their implementations must conform to the following requirements:</p>
+
+    <ul>
+      <li>At some future point in time, when there is no running execution context and the execution context stack is empty, the implementation must:
+        <ol>
+          <li>Push an execution context onto the execution context stack.</li>
+          <li>Perform any implementation-defined preparation steps.</li>
+          <li>Perform the abstract operation given by _pending_.[[Job]], passing the arguments _pending_.[[Arguments]].</li>
+          <li>Perform any implementation-defined cleanup steps.</li>
+          <li>Pop the previously-pushed execution context from the execution context stack.</li>
+        </ol>
+      </li>
+      <li>Only one Job may be actively undergoing evaluation at any point in time.</li>
+      <li>Once evaluation of a Job starts, it must run to completion before evaluation of any other Job starts.</li>
+      <li>The abstract operation must return a normal completion, implementing its own handling of errors.</li>
+    </ul>
 
-    <emu-clause id="sec-enqueuejob" aoid="EnqueueJob">
-      <h1>EnqueueJob ( _queueName_, _job_, _arguments_ )</h1>
-      <p>The EnqueueJob abstract operation requires three arguments: _queueName_, _job_, and _arguments_. It performs the following steps:</p>
+    <emu-clause id="sec-hostenqueuepromisejob" aoid="HostEnqueuePromiseJob">
+      <h1>HostEnqueuePromiseJob ( _pending_ )</h1>
+
+      <p>HostEnqueuePromiseJob is a host-defined abstract operation that schedules the PendingJob record _pending_ to be performed, at some future time. The PendingJob records used with this algorithm are intended to be related to the handling of Promises, or otherwise, to be scheduled with equal priority to Promise handling operations.</p>
+
+      <p>The implementation of HostEnqueuePromiseJob must conform to the requirements in <emu-xref href="#sec-jobs-and-job-queues"></emu-xref>. Additionally, Jobs must be scheduled in FIFO order, with Jobs running in the same order as the HostEnqueuePromiseJob invocations which scheduled them.</p>
+    </emu-clause>
+
+    <emu-clause id="sec-enqueuepromisejob" aoid="EnqueuePromiseJob">
+      <h1>EnqueuePromiseJob ( _job_, _arguments_ )</h1>
```

Comment:
What is the purpose of this indirection? It's less convenient as a host to have to unpack all of this from the _pending_ structure compared to the current direct call.

In other words, it'd be simpler if all EnqueuePromiseJob calls just called HostEnqueue(Promise)Job directly. And there's no need to speculatively gather up callerContext, callerRealm, and callerScriptOrModule; the HostEnqueue(Promise)Job implementation can take care of that itself if they're needed.

This would allow eliminating the PendingJob structure which is itself unnecessary complexity.

Reply:
These are very small editorial differences. Continuing to use the PendingJob data structure was a specific request of @erights when reviewing this patch, and easy to meet. It removes the need for the host to get these (if the host is comfortable referencing what JS passes it), and I think this could also be where we could pass the "entry Realm" that HTML needs, if we can get consensus on that layering change (c.f. https://github.com/whatwg/html/pull/5212).